File: LeetCodePractice/src/ReservoirSampling/LinkedListRandomNode.java
package ReservoirSampling;

import java.util.Random;

import LinkedList.ListNode;

/*
Given a singly linked list, return a random node's value from the linked list. Each node must have the same probability of being chosen.

Follow up:
What if the linked list is extremely large and its length is unknown to you? Could you solve this efficiently without using extra space?

Example:

// Init a singly linked list [1,2,3].
ListNode head = new ListNode(1);
head.next = new ListNode(2);
head.next.next = new ListNode(3);
Solution solution = new Solution(head);

// getRandom() should return either 1, 2, or 3 randomly. Each element should have equal probability of returning.
solution.getRandom();
*/
public class LinkedListRandomNode {
	Random random;
	ListNode head;

	/**
	 * @param head
	 *            The linked list's head. Note that the head is guaranteed to be
	 *            not null, so it contains at least one node.
	 */
	public LinkedListRandomNode(ListNode head) {
		random = new Random();
		this.head = head;
	}

	/** Returns a random node's value. */
	public int getRandom() {
		ListNode cur = head;
		int res = cur.val;
		int count = 1, probability;
		while (cur != null) {
			probability = random.nextInt(count);
			if (probability == 0) {
				res = cur.val;
			}
			cur = cur.next;
			count++;
		}
		return res;
	}
}
/**
 * Your Solution object will be instantiated and called as such: Solution obj =
 * new Solution(head); int param_1 = obj.getRandom();
 */

File: LeetCodePython/LinkedListRandomNode.py
"""
Given a singly linked list, return a random node's value from the linked list. Each node must have the same probability of being chosen.

Implement the Solution class:

Solution(ListNode head) Initializes the object with the head of the singly-linked list head.
int getRandom() Chooses a node randomly from the list and returns its value. All the nodes of the list should be equally likely to be chosen.

Example 1:
Input
["Solution", "getRandom", "getRandom", "getRandom", "getRandom", "getRandom"]
[[[1, 2, 3]], [], [], [], [], []]
Output
[null, 1, 3, 2, 2, 3]

Explanation
Solution solution = new Solution([1, 2, 3]);
solution.getRandom(); // return 1
solution.getRandom(); // return 3
solution.getRandom(); // return 2
solution.getRandom(); // return 2
solution.getRandom(); // return 3
// getRandom() should return either 1, 2, or 3 randomly. Each element should have equal probability of returning.

Constraints:
The number of nodes in the linked list will be in the range [1, 104].
-10^4 <= Node.val <= 10^4
At most 104 calls will be made to getRandom.

Follow up:
What if the linked list is extremely large and its length is unknown to you?
Could you solve this efficiently without using extra space?
"""
from random import random
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class LinkedListRandomNode:

    def __init__(self, head: Optional[ListNode]):
        self.head = head

    def getRandom(self) -> int:
        cur = self.head
        res, cnt, prob = -1, 0, 0
        while cur:
            prob = random.randint(0, cnt)
            if prob == 0:
                res = cur.val
            cur = cur.next
            cnt += 1
        return res





# Your Solution object will be instantiated and called as such:
# obj = Solution(head)
# param_1 = obj.getRandom()